blog/_posts/06-03-blogpost.md

```diff
@@ -0,0 +1,164 @@
+
+Once I had gotten this done, the speedups were generally quite high.
+
+![WIP BitVector Speedups All](BitVectorSpeedupsAll.png "BitVector Speedups All")
+
+![WIP BitVector Speedups](BitVectorSpeedups.png "BitVector Speedups")
+
+\<\<\<Section where I talk about the results in more detail\>\>\>
+
+## The DataMap Reference Capability
+This is how a DataMap used to be structured:
+
+![WIP DataMap Image](DataMap.png "DataMap")
+
+And this is how I have restructured it.
+
+![WIP New DataMap Image](NewDataMap.png "New DataMap")
+
+Note that the size is determined by the size of Ptr -- if we are in memory tracking mode, then a Ptr will contain more than just a ptr.  
+
+Unlike BitVector, there isn't one piece of extremely important code.  Instead, there are two, those being the functions Set and SetRef.
+
+```c++
+/// Set a variable by ID.
+template <typename  T> T  &  Set(size_t  id, const  T  &  value) {
+	if (memory.isRef(id)) {
+		memory.Destruct<T>(id);
+		memory.setRef(id, false);
+		new (memory.GetPtr<T>(id+1).Raw()) T(value);
+		return  memory.Get<T>(id);
+	} else {
+		return (Get<T>(id) = value);
+	}
+}
+
+/// Set a reference to a variable by ID.
+template <typename  T> T  &  SetRef(size_t  id, T  &  value) {
+	if (!memory.isRef(id)) {
+		memory.Destruct<T>(id);
+		memory.setRef(id, true);
+		new (memory.GetPtr<Ptr<T>>(id+1).Raw()) Ptr<T>(& value);
+	} else {
+		*memory.GetPtr<Ptr<T>>(id+1) = & value;
+	}
+	return  memory.Get<T>(id);
+}
+```
+
+The complex bit of this code is the "new" phrase.  It creates an instance of the object using the copy constructor at where the pointer to the object (which is actually a reference, but that doesn't matter since it's been destructed).  Turns out you need to do it this way or you end up with memory leaks.  That was probably the most difficult part of this project.
+
+Now, memory is (as you would expect) a MemoryImage, and MemoryImage::Get is 
+
+```c++
+/// Get proper references to an object represented in this image.
+template <typename  T> T  &  Get(size_t  pos) {
+	if (isRef(pos)) {
+		return **GetPtr<Ptr<T>>(pos+1);
+	}
+	return *GetPtr<T>(pos+1);
+}
+```
+
+As you can see, MemoryImage::Get properly indexes past the control byte into the actual memory, which means DataMap::Set and DataMap::SetRef don't have to.  This positioning is necessary, since DataMap isn't the only thing which calls MemoryImage::Get and MemoryImage::GetPtr.  I missed this, however, which led to quite a spectacular bug when I put the increment in MemoryImage::GetPtr instead.
```

Comment:
Again, I think wrapping stuff like `MemoryImage::Get` in tick marks will make it more readable